### PR TITLE
Endpoint UAMI

### PR DESCRIPTION
**Pull request by algattik**
Deploy a User-Assigned Managed Identity for the promptflow REST endpoint.

Previously, the endpoint had a System-Assigned Managed Identity (default setting) which had the following downsides:
- The first deployment failed, and we must manually assign IAM permissions to the MI before running the pipeline again.
- If the endpoint is deleted and recreated, a new MI is generated and the assignment must be done again

This change has one downside, that with the default setting permissions on Azure ML storage and container registry were automatically granted. With UAMI, these permissions must be manually granted (once) (this is reflected in updates to CI-CD.md).

Deployment tested here: https://github.com/algattik-mve/promptflow-demo/actions/runs/7127028330

### CODE REVIEW

File: CI-CD.md [hunk 1]
```diff
@@ -64,28 +62,48 @@ In Azure ML Studio, under `Prompt flow`, in the `Connections` tab, create the fo
...
+- Role: [`AcrPull`](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#acrpull)
+- Identity: the managed identity you created for the endpoint
+
+### Endpoint workspace secrets
+
+Grant the following IAM role assignments to [allow the endpoint to retrieve secrets](https://aka.ms/pf-deploy-identity):
 
-- Identity:  `Managed identity` -> ` Machine Learning online endpoint` -> `chat-with-patents`
 - Resource: your Azure ML workspace
+
```
Comment (roel4ez) [inline CI-CD.md]: nit: this breaks rendering the list correctly
```suggestion
```

File: CI-CD.md [hunk 2]
```diff
@@ -64,28 +62,48 @@ In Azure ML Studio, under `Prompt flow`, in the `Connections` tab, create the fo
...
 
-Grant the following RBAC role to [allow the endpoint to retrieve secrets](https://aka.ms/pf-deploy-identity):
+- Resource: the storage account for the Azure ML workspace
+- Role: [`Storage blob data reader`](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#storage-blob-data-reader)
+- Identity: the managed identity you created for the endpoint
+
+Second role assignment:
+
+- Resource: the Azure Container Registry (ACR) for the Azure ML workspace
+
```
Comment (roel4ez) [inline CI-CD.md]: nit: this breaks rendering the list correctly
```suggestion
```

File: CI-CD.md [hunk 3]
```diff
@@ -64,28 +62,48 @@ In Azure ML Studio, under `Prompt flow`, in the `Connections` tab, create the fo
...
+- Identity: the managed identity you created for CD
 
-## Grant Endpoint permissions resource access
+### Managed identity assignment
 
-After the CI/CD pipeline has run a first time, the endpoint will be created, but is missing the permissions to retrieve secrets to authenticate to Azure OpenAI and Azure AI Search.
+Grant the following IAM role assignments to allow  the pipeline to assign the managed identity to the deployed online endpoint:
 
-The pipeline contains a step to invoke the model, which will then fail with an error similar to:
+- Resource: the managed identity you created for the online endpoint
```
Comment (roel4ez) [inline CI-CD.md]: nit: this breaks rendering the list correctly
```suggestion
```

File: CI-CD.md [hunk 4]
```diff
@@ -64,28 +62,48 @@ In Azure ML Studio, under `Prompt flow`, in the `Connections` tab, create the fo
 
 The connections are used by the Azure ML endpoint deployed by the pipeline.
 
-## Grant Azure resource access
+## Grant Azure role assignments
```
Comment (roel4ez) [inline CI-CD.md:65]: nit: the following paragraph may be simplified into a table, e.g.:

| Action | Resource | Role | Identity | 
| --- | --- | --- | --- |
| allow the pipeline to deploy models | Azure ML Workspace | AzureML Data Scientist | `cd_mi` |
| etc... | | | |

**Review by roel4ez — APPROVED** [review]

File: .github/workflows/deploy-pf-online-endpoint-pipeline.yml [hunk 5]
```diff
@@ -42,18 +44,32 @@ jobs:
...
-        az ml online-endpoint update --file deployment/endpoint.yaml --name ${{env.ENDPOINT_NAME}}
+        az ml online-endpoint create $args || \
+        az ml online-endpoint update $args
+      env:
+        # https://learn.microsoft.com/en-us/azure/machine-learning/how-to-access-resources-from-endpoints-managed-identities?view=azureml-api-2&tabs=user-identity-cli#create-an-online-endpoint
+        args: >-
+          --file deployment/endpoint.yaml
+          --name ${{env.ENDPOINT_NAME}}
+          --set identity.user_assigned_identities[0].resource_id=${{env.ENDPOINT_IDENTITY_ARM_ID}}
     - name: Get deployment name
```
Comment (dburik) [inline .github/workflows/deploy-pf-online-endpoint-pipeline.yml:55]: nit: the name of this step is misleading now

**Review by dburik — COMMENTED** [review]
Looks good to me.

**Pull request merged**